packages/steps/src/utils/shell/spawn.ts

```diff
@@ -1,45 +1,161 @@
+const SPAWN_KILL_TIMEOUT_DEFAULT_ERROR_MESSAGE =
+  'Command was inactive for over ${minutes} minutes. Please evaluate if it is correct.';
+
+export async function spawnAsync(
   command: string,
   args: string[],
   allOptions: SpawnOptions = {
     stdio: 'inherit',
     cwd: process.cwd(),
   }
-): SpawnPromise<SpawnResult> {
-  const { logger, ...options } = allOptions;
-  const promise = spawnAsyncOriginal(command, args, options);
-  if (logger && promise.child) {
-    pipeSpawnOutput(logger, promise.child, options);
+): Promise<SpawnResult> {
+  const { logger, noLogsTimeout, ...options } = allOptions;
+  let spawnWarnTimeout: NodeJS.Timeout | undefined;
+  let spawnKillTimeout: NodeJS.Timeout | undefined;
+  let spawnTimedOut: boolean = false;
+  let spawnResult: SpawnResult;
+
+  function setCommandSpawnTimeouts(
+    noLogsTimeout: NoLogsTimeoutOptions,
+    logger: bunyan,
+    spawnPromise: SpawnPromise<SpawnResult>
+  ): void {
+    if (noLogsTimeout.warn) {
+      const warnTimeoutMinutes =
+        noLogsTimeout.warn.timeoutMinutes ?? SPAWN_WARN_TIMEOUT_DEFAULT_MINUTES;
+      spawnWarnTimeout = setTimeout(
+        () => {
+          logger.warn(
+            noLogsTimeout.warn?.message ??
+              SPAWN_WARN_TIMEOUT_DEFAULT_MESSAGE.replace(
+                '${minutes}',
+                warnTimeoutMinutes.toString()
+              )
+          );
+        },
+        warnTimeoutMinutes * 60 * 1000
+      );
+    }
+
+    if (noLogsTimeout.kill) {
+      const killTimeoutMinutes =
+        noLogsTimeout.kill.timeoutMinutes ?? SPAWN_KILL_TIMEOUT_DEFAULT_MINUTES;
+      spawnKillTimeout = setTimeout(
+        async () => {
+          spawnTimedOut = true;
+          logger.error(
+            noLogsTimeout.kill?.message ??
+              SPAWN_KILL_TIMEOUT_DEFAULT_MESSAGE.replace(
+                '${minutes}',
+                killTimeoutMinutes.toString()
+              )
+          );
+          const ppid = nullthrows(spawnPromise.child.pid);
+          const pids = await getParentAndDescendantProcessPidsAsync(ppid);
+          pids.forEach((pid) => {
+            process.kill(pid);
+          });
+        },
+        killTimeoutMinutes * 60 * 1000
+      );
+    }
+  }
+
+  try {
+    const spawnPromise = spawnAsyncOriginal(command, args, options);
+    if (logger && spawnPromise.child) {
+      if (noLogsTimeout) {
+        const optionsWithCallback = {
+          ...options,
+          infoCallbackFn: () => {
+            if (spawnWarnTimeout) {
+              spawnWarnTimeout.refresh();
+            }
+            if (spawnKillTimeout) {
+              spawnKillTimeout.refresh();
+            }
+          },
+        };
+        pipeSpawnOutput(logger, spawnPromise.child, optionsWithCallback);
+        setCommandSpawnTimeouts(noLogsTimeout, logger, spawnPromise);
+      } else {
+        pipeSpawnOutput(logger, spawnPromise.child, options);
+      }
+    }
+    spawnResult = await spawnPromise;
```

Comment:
Hmm, I think in some places we don't `await` the `SpawnPromise` by design. E.g. [here](https://github.com/expo/eas-build/blob/e901e53419a181e4d1bdfcc3bd8ac4a952f0bbac/packages/build-tools/src/steps/functions/startAndroidEmulator.ts#L67-L80).

Instead of `try-catch` and `await`ing the `spawnPromise`, could we attach the `catch` as `spawnPromise.catch(err => ...)`?